File: TP4/Game.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#ifdef _WIN32
#define strcasecmp _stricmp
#endif

#define MAX_LINE 8192
#define MAX_FIELD 2048
#define MAX_LIST 100
#define MAX_GAMES 50000

typedef struct {
    char **items;
    int count;
} StringList;

typedef struct {
    int id;
    char *name;
    char releaseDate[11]; // dd/MM/yyyy
    int estimatedOwners;
    float price;
    StringList supportedLanguages;
    int metacriticScore;
    float userScore;
    int achievements;
    StringList publishers;
    StringList developers;
    StringList categories;
    StringList genres;
    StringList tags;
} Game;

// ===== Helper Functions =====
static char* my_strndup(const char* s, size_t n) {
    char* p = (char*)malloc(n + 1);
    if (!p) return NULL;
    memcpy(p, s, n);
    p[n] = '\0';
    return p;
}

static int month_from_abbrev(const char* m) {
    char buf[4];
    if (!m || strlen(m) < 3) return 0;
    buf[0] = tolower((unsigned char)m[0]);
    buf[1] = tolower((unsigned char)m[1]);
    buf[2] = tolower((unsigned char)m[2]);
    buf[3] = '\0';
    if (!strcmp(buf,"jan")) return 1;
    if (!strcmp(buf,"feb")) return 2;
    if (!strcmp(buf,"mar")) return 3;
    if (!strcmp(buf,"apr")) return 4;
    if (!strcmp(buf,"may")) return 5;
    if (!strcmp(buf,"jun")) return 6;
    if (!strcmp(buf,"jul")) return 7;
    if (!strcmp(buf,"aug")) return 8;
    if (!strcmp(buf,"sep")) return 9;
    if (!strcmp(buf,"oct")) return 10;
    if (!strcmp(buf,"nov")) return 11;
    if (!strcmp(buf,"dec")) return 12;
    return 0;
}

// ===== String List Functions =====
void initList(StringList *list) {
    list->items = malloc(MAX_LIST * sizeof(char*));
    list->count = 0;
}

void addToList(StringList *list, const char *item) {
    if (list->count < MAX_LIST) {
        list->items[list->count] = strdup(item);
        list->count++;
    }
}

void freeList(StringList *list) {
    for (int i = 0; i < list->count; i++) {
        free(list->items[i]);
    }
    free(list->items);
}

// ===== Parsing Helper Functions =====
char* cleanQuotes(char *s) {
    if (!s) return strdup("");
    while (isspace(*s)) s++;
    int len = strlen(s);
    while (len > 0 && isspace(s[len-1])) len--;
    if (len >= 2 && ((s[0] == '"' && s[len-1] == '"') || (s[0] == '\'' && s[len-1] == '\''))) {
        s++; len -= 2;
    }
    char *result = malloc(len + 1);
    strncpy(result, s, len);
    result[len] = '\0';
    return result;
}

int parseIntOrDefault(char *s, int def) {
    char *clean = cleanQuotes(s);
    if (strlen(clean) == 0) {
        free(clean);
        return def;
    }
    int val = atoi(clean);
    free(clean);
    return val;
}

float parseFloatOrDefault(char *s, float def) {
    char *clean = cleanQuotes(s);
    if (strlen(clean) == 0) {
        free(clean);
        return def;
    }
    float val = atof(clean);
    free(clean);
    return val;
}

int parseOwners(char *s) {
    char *clean = cleanQuotes(s);
    char digits[256] = "";
    int idx = 0;
    for (int i = 0; clean[i] && idx < 255; i++) {
        if (isdigit(clean[i])) digits[idx++] = clean[i];
    }
    digits[idx] = '\0';
    free(clean);
    return strlen(digits) > 0 ? atoi(digits) : 0;
}

float parsePrice(char *s) {
    char *clean = cleanQuotes(s);
    if (strlen(clean) == 0 || strcasecmp(clean, "Free to Play") == 0) {
        free(clean);
        return 0.0f;
    }
    char *p = clean;
    while (*p) {
        if (*p == '$' || *p == ',') memmove(p, p+1, strlen(p));
        else p++;
    }
    float val = atof(clean);
    free(clean);
    return val;
}

float parseUserScore(char *s) {
    char *clean = cleanQuotes(s);
    if (strlen(clean) == 0 || strcasecmp(clean, "tbd") == 0) {
        free(clean);
        return -1.0f;
    }
    float val = atof(clean);
    free(clean);
    return val;
}

void normalizeDate(char *s, char *out) {
    strcpy(out, "01/01/0001");

    char *clean = cleanQuotes(s);
    if (!clean || clean[0] == '\0') {
        if (clean) free(clean);
        return;
    }

    // 1) Se for apenas ano
    int onlyYear = 1;
    for (int i = 0; clean[i]; i++) {
        if (!isdigit((unsigned char)clean[i]) && !isspace((unsigned char)clean[i])) {
            onlyYear = 0; break;
        }
    }
    if (onlyYear) {
        int y = 0; int cnt = 0;
        for (int i = 0; clean[i]; i++) {
            if (isdigit((unsigned char)clean[i])) {
                y = y*10 + (clean[i]-'0');
                cnt++;
                if (cnt == 4) break;
            }
        }
        if (cnt >= 4 && y > 0) {
            sprintf(out, "01/01/%04d", y);
            free(clean);
            return;
        }
    }

    // 2) Tentar "MMM d, yyyy" | "MMM dd, yyyy"
    {
        char mmm[16]; int d=0, y=0;
        if (sscanf(clean, " %15s %d , %d", mmm, &d, &y) == 3) {
            int m = month_from_abbrev(mmm);
            if (m >= 1 && m <= 12 && d >= 1 && d <= 31 && y >= 1) {
                sprintf(out, "%02d/%02d/%04d", d, m, y);
                free(clean);
                return;
            }
        }
    }
    // 3) Tentar "d MMM, yyyy" | "dd MMM, yyyy"
    {
        char mmm[16]; int d=0, y=0;
        if (sscanf(clean, " %d %15s , %d", &d, mmm, &y) == 3) {
            int m = month_from_abbrev(mmm);
            if (m >= 1 && m <= 12 && d >= 1 && d <= 31 && y >= 1) {
                sprintf(out, "%02d/%02d/%04d", d, m, y);
                free(clean);
                return;
            }
        }
    }
    // 4) Tentar "MMM yyyy" (assume dia 1)
    {
        char mmm[16]; int y=0;
        if (sscanf(clean, " %15s %d", mmm, &y) == 2) {
            int m = month_from_abbrev(mmm);
            if (m >= 1 && m <= 12 && y >= 1) {
                sprintf(out, "%02d/%02d/%04d", 1, m, y);
                free(clean);
                return;
            }
        }
    }
    // 5) Tentar pegar dígitos "d m yyyy" ou "m yyyy"
    {
        char digits[128] = {0};
        int j = 0;
        for (int i = 0; clean[i] && j < 120; i++) {
            if (isdigit((unsigned char)clean[i])) digits[j++] = clean[i];
            else digits[j++] = ' ';
        }
        digits[j] = '\0';
        char norm[128] = {0};
        int k = 0, inSpace = 0;
        for (int i = 0; digits[i] && k < 120; i++) {
            if (isspace((unsigned char)digits[i])) {
                if (!inSpace) { norm[k++] = ' '; inSpace = 1; }
            } else {
                norm[k++] = digits[i];
                inSpace = 0;
            }
        }
        norm[k] = '\0';

        int a=0,b=0,c=0;
        if (sscanf(norm, " %d %d %d", &a, &b, &c) == 3) {
            if (a>=1 && a<=31 && b>=1 && b<=12 && c>=1) {
                sprintf(out, "%02d/%02d/%04d", a, b, c);
                free(clean);
                return;
            }
        }
        if (sscanf(norm, " %d %d", &a, &b) == 2) {
            if (a>=1 && a<=12 && b>=1) {
                sprintf(out, "%02d/%02d/%04d", 1, a, b);
                free(clean);
                return;
            }
        }
    }

    free(clean);
    strcpy(out, "01/01/0001");
}

void parseListBracketed(char *s, StringList *list) {
    if (!s || strlen(s) == 0) return;
    char *start = strchr(s, '[');
    char *end = strrchr(s, ']');
    if (!start || !end || end <= start) return;
    
    char *inner = my_strndup(start + 1, (size_t)(end - start - 1));
    char *token = strtok(inner, ",");
    while (token) {
        char *clean = cleanQuotes(token);
        if (strlen(clean) > 0) addToList(list, clean);
        free(clean);
        token = strtok(NULL, ",");
    }
    free(inner);
}

void parseListComma(char *s, StringList *list) {
    if (!s || strlen(s) == 0) return;
    if (strchr(s, '[') && strchr(s, ']')) {
        parseListBracketed(s, list);
        return;
    }
    char *copy = strdup(s);
    char *token = strtok(copy, ",");
    while (token) {
        char *clean = cleanQuotes(token);
        if (strlen(clean) > 0) addToList(list, clean);
        free(clean);
        token = strtok(NULL, ",");
    }
    free(copy);
}

void formatList(StringList *list, char *out) {
    if (list->count == 0) {
        strcpy(out, "[]");
        return;
    }
    strcpy(out, "[");
    for (int i = 0; i < list->count; i++) {
        strcat(out, list->items[i]);
        if (i < list->count - 1) strcat(out, ", ");
    }
    strcat(out, "]");
}

void formatFloat(float v, char *out) {
    sprintf(out, "%.10g", v);
}

// ===== CSV Parsing =====
int splitCSV(char *line, char fields[][MAX_FIELD]) {
    int fieldCount = 0;
    char cur[MAX_FIELD] = "";
    int curIdx = 0;
    int inQuotes = 0;
    int bracketLevel = 0;
    
    for (int i = 0; line[i]; i++) {
        char c = line[i];
        if (c == '"') {
            inQuotes = !inQuotes;
            cur[curIdx++] = c;
        } else if (c == '[' && !inQuotes) {
            bracketLevel++;
            cur[curIdx++] = c;
        } else if (c == ']' && !inQuotes && bracketLevel > 0) {
            bracketLevel--;
            cur[curIdx++] = c;
        } else if (c == ',' && !inQuotes && bracketLevel == 0) {
            cur[curIdx] = '\0';
            strcpy(fields[fieldCount++], cur);
            curIdx = 0;
        } else {
            cur[curIdx++] = c;
        }
    }
    cur[curIdx] = '\0';
    strcpy(fields[fieldCount++], cur);
    return fieldCount;
}

// ===== Game Functions =====
void initGame(Game *g) {
    g->id = 0;
    g->name = NULL;
    strcpy(g->releaseDate, "01/01/0001");
    g->estimatedOwners = 0;
    g->price = 0.0f;
    g->metacriticScore = -1;
    g->userScore = -1.0f;
    g->achievements = 0;
    initList(&g->supportedLanguages);
    initList(&g->publishers);
    initList(&g->developers);
    initList(&g->categories);
    initList(&g->genres);
    initList(&g->tags);
}

void freeGame(Game *g) {
    if (g->name) free(g->name);
    freeList(&g->supportedLanguages);
    freeList(&g->publishers);
    freeList(&g->developers);
    freeList(&g->categories);
    freeList(&g->genres);
    freeList(&g->tags);
}

void printGame(Game *g) {
    char buf[MAX_LINE * 2];
    char temp[MAX_LINE];
    
    sprintf(buf, "=> %d ## %s ## %s ## %d ## ", g->id, g->name, g->releaseDate, g->estimatedOwners);
    formatFloat(g->price, temp);
    strcat(buf, temp);
    strcat(buf, " ## ");
    formatList(&g->supportedLanguages, temp);
    strcat(buf, temp);
    sprintf(temp, " ## %d ## ", g->metacriticScore);
    strcat(buf, temp);
    formatFloat(g->userScore, temp);
    strcat(buf, temp);
    sprintf(temp, " ## %d ## ", g->achievements);
    strcat(buf, temp);
    formatList(&g->publishers, temp);
    strcat(buf, temp);
    strcat(buf, " ## ");
    formatList(&g->developers, temp);
    strcat(buf, temp);
    strcat(buf, " ## ");
    formatList(&g->categories, temp);
    strcat(buf, temp);
    strcat(buf, " ## ");
    formatList(&g->genres, temp);
    strcat(buf, temp);
    strcat(buf, " ## ");
    formatList(&g->tags, temp);
    strcat(buf, temp);
    strcat(buf, " ##");
    
    printf("%s\n", buf);
}

// ===== Main =====
int main() {
    Game *games = malloc(MAX_GAMES * sizeof(Game));
    int gameCount = 0;
    
    FILE *fp = fopen("/tmp/games.csv", "r");
    if (!fp) {
        fprintf(stderr, "Falha ao abrir games.csv\n");
        return 1;
    }
    
    char line[MAX_LINE];
    fgets(line, MAX_LINE, fp); // skip header
    
    while (fgets(line, MAX_LINE, fp)) {
        line[strcspn(line, "\n")] = 0;
        if (strlen(line) == 0) continue;
        
        char fields[20][MAX_FIELD];
        int fieldCount = splitCSV(line, fields);
        if (fieldCount < 14) continue;
        
        Game g;
        initGame(&g);
        g.id = parseIntOrDefault(fields[0], 0);
        g.name = cleanQuotes(fields[1]);
        normalizeDate(fields[2], g.releaseDate);
        g.estimatedOwners = parseOwners(fields[3]);
        g.price = parsePrice(fields[4]);
        parseListBracketed(fields[5], &g.supportedLanguages);
        g.metacriticScore = parseIntOrDefault(fields[6], -1);
        g.userScore = parseUserScore(fields[7]);
        g.achievements = parseIntOrDefault(fields[8], 0);
        parseListComma(fields[9], &g.publishers);
        parseListComma(fields[10], &g.developers);
        parseListComma(fields[11], &g.categories);
        parseListComma(fields[12], &g.genres);
        parseListComma(fields[13], &g.tags);
        
        if (g.id != 0) games[gameCount++] = g;
    }
    fclose(fp);
    
    char input[100];
    while (fgets(input, 100, stdin)) {
        input[strcspn(input, "\n")] = 0;
        if (strcmp(input, "FIM") == 0) break;
        if (strlen(input) == 0) continue;
        
        int id = atoi(input);
        for (int i = 0; i < gameCount; i++) {
            if (games[i].id == id) {
                printGame(&games[i]);
                break;
            }
        }
    }
    
    for (int i = 0; i < gameCount; i++) {
        freeGame(&games[i]);
    }
    free(games);
    
    return 0;
}

File: TP4/Game.java
import java.io.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.*;

public class Game {
    private int id;
    private String name;
    private String releaseDate; // dd/MM/yyyy
    private int estimatedOwners;
    private float price;
    private List<String> supportedLanguages;
    private int metacriticScore; // vazio -> -1
    private float userScore; // vazio ou tbd -> -1.0
    private int achievements; // vazio -> 0
    private List<String> publishers;
    private List<String> developers;
    private List<String> categories;
    private List<String> genres;
    private List<String> tags;

    public int getId() {
        return id;
    }

    // ===== Helpers de parsing =====
    private static String cleanQuotes(String s) {
        if (s == null)
            return "";
        s = s.trim();
        if ((s.startsWith("\"") && s.endsWith("\"")) || (s.startsWith("'") && s.endsWith("'"))) {
            if (s.length() >= 2)
                s = s.substring(1, s.length() - 1);
        }
        return s.trim();
    }

    private static int parseIntOrDefault(String s, int def) {
        s = cleanQuotes(s);
        if (s.isEmpty())
            return def;
        try {
            return Integer.parseInt(s);
        } catch (Exception e) {
            return def;
        }
    }

    private static float parseFloatOrDefault(String s, float def) {
        s = cleanQuotes(s);
        if (s.isEmpty())
            return def;
        try {
            return Float.parseFloat(s);
        } catch (Exception e) {
            return def;
        }
    }

    private static int parseOwners(String s) {
        s = cleanQuotes(s);
        if (s.isEmpty())
            return 0;
        String digits = s.replaceAll("[^0-9]", "");
        if (digits.isEmpty())
            return 0;
        try {
            return Integer.parseInt(digits);
        } catch (Exception e) {
            return 0;
        }
    }

    private static float parsePrice(String s) {
        s = cleanQuotes(s);
        if (s.isEmpty())
            return 0.0f;
        if (s.equalsIgnoreCase("Free to Play"))
            return 0.0f;
        s = s.replace("$", "").replace(",", "").trim();
        return parseFloatOrDefault(s, 0.0f);
    }

    private static float parseUserScore(String s) {
        s = cleanQuotes(s);
        if (s.isEmpty())
            return -1.0f;
        if (s.equalsIgnoreCase("tbd"))
            return -1.0f;
        return parseFloatOrDefault(s, -1.0f);
    }

    private static String normalizeDate(String s) {
        s = cleanQuotes(s);
        if (s.isEmpty())
            return "01/01/0001";
        String[] inFormats = {
                "MMM d, yyyy", "MMM dd, yyyy",
                "d MMM, yyyy", "dd MMM, yyyy",
                "MMM yyyy",
                "yyyy"
        };
        for (String fmt : inFormats) {
            try {
                SimpleDateFormat sdf = new SimpleDateFormat(fmt, Locale.ENGLISH);
                sdf.setLenient(false);
                Date d = sdf.parse(s);
                return new SimpleDateFormat("dd/MM/yyyy").format(d);
            } catch (ParseException ignore) {
            }
        }
        if (s.matches("\\d{4}")) {
            int y = Integer.parseInt(s);
            return String.format("%02d/%02d/%04d", 1, 1, y);
        }
        String digits = s.replaceAll("[^0-9]", " ").trim().replaceAll("\\s+", " ");
        String[] parts = digits.isEmpty() ? new String[0] : digits.split(" ");
        if (parts.length == 3) {
            try {
                int d = Integer.parseInt(parts[0]);
                int m = Integer.parseInt(parts[1]);
                int y = Integer.parseInt(parts[2]);
                return String.format("%02d/%02d/%04d", d, m, y);
            } catch (Exception ignore) {
            }
        } else if (parts.length == 2) {
            try {
                int m = Integer.parseInt(parts[0]);
                int y = Integer.parseInt(parts[1]);
                return String.format("%02d/%02d/%04d", 1, m, y);
            } catch (Exception ignore) {
            }
        }
        return "01/01/0001";
    }

    private static List<String> parseListBracketed(String s) {
        List<String> out = new ArrayList<>();
        s = (s == null) ? "" : s.trim();
        if (s.isEmpty())
            return out;
        int l = s.indexOf('['), r = s.lastIndexOf(']');
        String inner = (l != -1 && r != -1 && r > l) ? s.substring(l + 1, r) : s;
        String[] parts = inner.split(",");
        LinkedHashSet<String> uniq = new LinkedHashSet<>();
        for (String p : parts) {
            String item = cleanQuotes(p);
            if (!item.isEmpty())
                uniq.add(item);
        }
        out.addAll(uniq);
        return out;
    }

    private static List<String> parseListComma(String s) {
        if (s == null)
            return new ArrayList<>();
        s = s.trim();
        if (s.isEmpty())
            return new ArrayList<>();
        if (s.contains("[") && s.contains("]"))
            return parseListBracketed(s);
        String[] parts = s.split(",");
        LinkedHashSet<String> uniq = new LinkedHashSet<>();
        for (String p : parts) {
            String item = cleanQuotes(p);
            if (!item.isEmpty())
                uniq.add(item);
        }
        return new ArrayList<>(uniq);
    }

    private static String formatList(List<String> list) {
        if (list == null || list.isEmpty())
            return "[]";
        return "[" + String.join(", ", list) + "]";
    }

    private static String formatFloatForOutput(float v) {
        String s = Float.toString(v);
        if (s.endsWith(".0")) {
            if (v == 0.0f || v == -1.0f)
                return s;
            return s.substring(0, s.length() - 2);
        }
        s = s.replaceAll("(\\.\\d*?[1-9])0+$", "$1");
        return s;
    }

    public String toJudgeLine() {
        return "=> " + id + " ## " + name + " ## " + releaseDate + " ## " + estimatedOwners + " ## "
                + formatFloatForOutput(price) + " ## " + formatList(supportedLanguages) + " ## "
                + metacriticScore + " ## " + formatFloatForOutput(userScore) + " ## " + achievements + " ## "
                + formatList(publishers) + " ## " + formatList(developers) + " ## " + formatList(categories)
                + " ## " + formatList(genres) + " ## " + formatList(tags) + " ##";
    }

    // ===== CSV: vírgula como separador, respeitando aspas e colchetes =====
    private static String[] splitCSV(String line) {
        List<String> fields = new ArrayList<>();
        StringBuilder cur = new StringBuilder();
        boolean inQuotes = false;
        int bracketLevel = 0;

        for (int i = 0; i < line.length(); i++) {
            char c = line.charAt(i);
            if (c == '"') {
                inQuotes = !inQuotes;
                cur.append(c);
            } else if (c == '[' && !inQuotes) {
                bracketLevel++;
                cur.append(c);
            } else if (c == ']' && !inQuotes && bracketLevel > 0) {
                bracketLevel--;
                cur.append(c);
            } else if (c == ',' && !inQuotes && bracketLevel == 0) {
                fields.add(cur.toString());
                cur.setLength(0);
            } else {
                cur.append(c);
            }
        }
        fields.add(cur.toString());
        return fields.toArray(new String[0]);
    }

    private static Map<Integer, Game> loadAllGames(String fileName) throws IOException {
        Map<Integer, Game> map = new HashMap<>();
        try (BufferedReader br = new BufferedReader(new FileReader(fileName))) {
            String header = br.readLine(); // cabeçalho (vírgula)
            if (header == null)
                return map;

            // índices fixos com base no seu header
            // AppID,Name,Release date,Estimated owners,Price,Supported languages,Metacritic
            // score,User score,Achievements,Publishers,Developers,Categories,Genres,Tags
            String line;
            while ((line = br.readLine()) != null) {
                if (line.isEmpty())
                    continue;
                String[] f = splitCSV(line);
                if (f.length < 14)
                    continue;

                Game g = new Game();
                g.id = parseIntOrDefault(f[0], 0);
                g.name = cleanQuotes(f[1]);
                g.releaseDate = normalizeDate(f[2]);
                g.estimatedOwners = parseOwners(f[3]);
                g.price = parsePrice(f[4]);
                g.supportedLanguages = parseListBracketed(f[5]);
                g.metacriticScore = parseIntOrDefault(f[6], -1);
                g.userScore = parseUserScore(f[7]);
                g.achievements = parseIntOrDefault(f[8], 0);
                g.publishers = parseListComma(f[9]);
                g.developers = parseListComma(f[10]);
                g.categories = parseListComma(f[11]); // no seu trecho, vem como "Single-player,Steam Achievements,..."
                                                      // entre aspas
                g.genres = parseListComma(f[12]); // idem
                g.tags = parseListComma(f[13]); // idem

                if (g.id != 0)
                    map.put(g.id, g);
            }
        }
        return map;
    }

    // ===== Main =====
    public static void main(String[] args) {
        Map<Integer, Game> byId;
        try {
            byId = loadAllGames("/tmp/games.csv");
        } catch (IOException e) {
            System.err.println("Falha ao abrir games.csv: " + e.getMessage());
            return;
        }

        Scanner sc = new Scanner(System.in);
        while (sc.hasNextLine()) {
            String s = sc.nextLine().trim();
            if (s.equals("FIM"))
                break;
            if (s.isEmpty())
                continue;
            try {
                int id = Integer.parseInt(s);
                Game g = byId.get(id);
                if (g != null) {
                    System.out.println(g.toJudgeLine());
                }
            } catch (NumberFormatException ignore) {
            }
        }
        sc.close();
    }
}